horovod/common/tensor_queue.cc

```diff
@@ -69,26 +69,80 @@ TensorQueue::GetTensorDataForAutotuner(const ResponseList& response_list,
 // Parse tensor names from response and generate a vector of corresponding
 // tensor entries.
 void TensorQueue::GetTensorEntriesFromResponse(
-    Response& response, std::vector<TensorTableEntry>& entries) {
+    Response& response, std::vector<TensorTableEntry>& entries,
+    int rank, bool joined, int join_device) {
   // Reserve to save re-allocation costs, as we know the size before.
   entries.reserve(response.tensor_names().size());
   {
     // Lock on the tensor table.
     std::lock_guard<std::mutex> guard(mutex_);
     for (auto& name : response.tensor_names()) {
-      // We should never fail at finding this key in the tensor table.
-      auto iter = tensor_table_.find(name);
-      assert(iter != tensor_table_.end());
-
       assert(response.response_type() == Response::ALLREDUCE ||
              response.response_type() == Response::ALLGATHER ||
              response.response_type() == Response::BROADCAST ||
              response.response_type() == Response::ERROR);
 
-      entries.push_back(std::move(iter->second));
-
-      // Clear the tensor table of this tensor.
-      tensor_table_.erase(iter);
+      if (!joined) {
+        // We should never fail at finding this key in the tensor table.
+        auto iter = tensor_table_.find(name);
+        assert(iter != tensor_table_.end());
+
+        entries.push_back(std::move(iter->second));
+
+        // Clear the tensor table of this tensor.
+        tensor_table_.erase(iter);
+      } else {
+        TensorTableEntry entry;
+        switch (response.tensor_type()) {
+        case HOROVOD_UINT8:
+          entry.tensor = std::make_shared<JoinTensor<HOROVOD_UINT8, uint8_t>>(
+              join_device, response.tensor_sizes()[0]);
+          break;
+        case HOROVOD_INT8:
+          entry.tensor = std::make_shared<JoinTensor<HOROVOD_INT8, int8_t>>(
+              join_device, response.tensor_sizes()[0]);
+          break;
+        case HOROVOD_UINT16:
+          entry.tensor = std::make_shared<JoinTensor<HOROVOD_UINT16, uint16_t>>(
+              join_device, response.tensor_sizes()[0]);
+          break;
+        case HOROVOD_INT16:
+          entry.tensor = std::make_shared<JoinTensor<HOROVOD_INT16, int16_t>>(
+              join_device, response.tensor_sizes()[0]);
+          break;
+        case HOROVOD_INT32:
+          entry.tensor = std::make_shared<JoinTensor<HOROVOD_INT32, int32_t>>(
+              join_device, response.tensor_sizes()[0]);
+          break;
+        case HOROVOD_INT64:
+          entry.tensor = std::make_shared<JoinTensor<HOROVOD_INT64, int64_t>>(
+              join_device, response.tensor_sizes()[0]);
+          break;
+        case HOROVOD_FLOAT16:
+          entry.tensor = std::make_shared<JoinTensor<HOROVOD_FLOAT16, float>>(
```

Comment:
@tgaddair Should it be eventually a switch statement on framework, then calling framework-specific allocator? This doesn't seem right, because now Join op code needs to know about every framework.
If `float` doesn't work for fp16 (it works in my tests, but maybe I need better tests), maybe some int variant will work? It only needs to work for value 0...

Reply:
There's no need to use a switch statement here.  If you implement the allocation of the correct tensor within [OpContext](https://github.com/horovod/horovod/blob/master/horovod/common/common.h#L206), then polymorphism will take care the rest.

Reply:
Hi @tgaddair,
So I'm planning to implement it this way. Any suggestions?

In `GetTensorEntriesFromResponse` to create temporary zero tensor, call `AllocatePersistent` of the `JOIN_TENSOR_NAME` context. `AllocatePersistent` will automatically call framework-specific implementation.

`AllocatePersistent` needs size in bytes, that can be computed from size in elements using  `GetTypeSize()` https://github.com/horovod/horovod/blob/c5b53eca9686ba296ba3a8b3a7e25dda25065eea/horovod/common/gloo/gloo_controller.cc#L94

The newly allocated buffer needs to be filled with zeros. Not really sure how to do it best, especially for fp16.

Or do you think reusing `AllocatePersistent` for this is not a good idea?

Reply:
Hey @kit1980, there are two limitations to `AllocatePersistent`, one is (as you mentioned) it returns an empty tensor, not a zeroed-tensor, and the other is that it allocates a tensor of data type int8.  What I think would be better would be to have a new method alongside `AllocatePersistent` like `AllocateZeros` that creates a zeroed-tensor of a given data type.  

So:

```
AllocateZeros(int device, int64_t size, horovod::common::DataType dtype)
```

TensorFlow and PyTorch both have C++ APIs for creating zero tensors and/or zeroing out tensors that can be used.